2015.2: backport overhaul of LazyLoader from #20274

jacksontj opened the pull request.

jacksontj committed:
Some major hacking of the loader, don't really like it... but i guess i'll keep it around for a while

Conflicts:
	salt/loader.py

jacksontj committed:
Some more cleanup to grains outputter, closer to working

Conflicts:
	salt/output/grains.py

jacksontj committed:
Migrate everything over to NewLazyLoader

Conflicts:
	salt/loader.py

jacksontj committed:
Fix grains caching

jacksontj committed:
Handle disk better

Since we are now loading the "squashfs" into memory, we need to handle the inevitable case where the disk content changes. This will re-load from disk up to once per get when a file is missing or we loaded everything else.

jacksontj committed:
Re-add pyx support

jacksontj committed:
pylint

jacksontj committed:
Catch case where 'return' not in opts, or other ways to get an empty returner (as it will just fail anyways)

jacksontj committed:
Add support for multiple types of functions by namespacing out the dep dict

Conflicts:
	salt/utils/decorators/__init__.py

jacksontj committed:
Add depends support to new lazyloader

jacksontj committed:
Fix type error

jacksontj committed:
Use clear() to reload

jacksontj committed:
Move to clear() for new loader

jacksontj committed:
Remove majority of *old* loader code

Conflicts:
	salt/loader.py

jacksontj committed:
Add support for non-virtual loader dicts and remove old loader class

Conflicts:
	salt/loader.py

jacksontj committed:
Support proxy modules, and misc pylint fixup

jacksontj committed:
Pass in "kind" to cleanup in depends decorator instead of guessing

Conflicts:
	salt/utils/decorators/__init__.py

jacksontj committed:
Reload submodules after initial load

jacksontj committed:
whitespace pylint cleanup

jacksontj committed:
More pylint fixes

jacksontj committed:
Fix call() in loader

jacksontj committed:
Fix import namespacing

jacksontj committed:
Pack modules to runners

jacksontj committed:
Misc other loader globals fixes

jacksontj committed:
Stop caching grains globally in memory.

We can add this later, but we have to determine what conditions go into unique-ing the grains (just mimic disk cache)

jacksontj committed:
Pylint fix

jacksontj committed:
pylint some more

jacksontj committed:
Fix namespacing of modules in dependency dict

jacksontj committed:
Fix pyobjects renderer to use new loader

jacksontj committed:
Make LazyLoader's singleton-ness optional

jacksontj committed:
Change default to false, since everyone seems to pass them in

jacksontj committed:
Migrate to new loader

jacksontj committed:
Make the pyobjects loader ignore virtual (it did before)

jacksontj committed:
Proper precedence of file extensions

jacksontj committed:
Pylint

jacksontj committed:
Rename func, since it overlaps (works because of ordering, but still...)

jacksontj committed:
__salt__ is not a lazydict object, not a dict.

In addition, since _yaml_result_unicode_to_utf8 is recursive, there is no need for the depth under the OrderedDict section

jacksontj committed:
Rename NewLazyLoader to LazyLoader (since the old one was deleted)

jacksontj committed:
Remove grains outputter

Its been broken for ~8 months (or more), and its been falling back to nested. Since nested does (IMO) a better job displaying the data and is generic, there is no reason to get this working-- esp since we've already "deprecated" it, by it not working

Conflicts:

	salt/output/grains.py

Conflicts:
	salt/output/grains.py

jacksontj committed:
Remove test as well

Conflicts:

	tests/unit/output/grains.py

Conflicts:
	tests/unit/output/grains.py

jacksontj committed:
Conditionally inject context (if no one else has)

jacksontj committed:
Attempt dict access

Apparently SSH FunctionWraper breaks when you do just the "in" comparator

jacksontj committed:
Inject globals after verifying the function-- because it might not exist!

Conflicts:
	salt/client/mixins.py

jacksontj committed:
Move "pack" over to a dict

Since it doesn't make sense to allow for duplicate "name" pack items, we should just use a dict. This also makes sharing __context__ between all the plugins *significantly* cleaner

jacksontj committed:
Remove unecessary format, a string is always a string

jacksontj committed:
Cleanup FunctionWrapper

jacksontj committed:
Misc cleanup

jacksontj committed:
Fix process_virtual error_reasons

Never used, so I fixed it up to not be a list-- since you can only have one. We will also log this as an error, for now since the module is asking for an error

jacksontj committed:
remove old deprecated mechanism

jacksontj committed:
Check that __context__ is always present, since it now is (properly)

jacksontj committed:
If you have the module loaded, lets not require another hit to disk

jacksontj committed:
Minor performance improvement

jacksontj committed:
Since clear() removes all modules, reloads are unecessary

jacksontj committed:
Add loader tests

jacksontj committed:
More loader tests

As well as an optimization on clear()

jacksontj committed:
Remove some old unused code

jacksontj committed:
Add test for __func__alias__

jacksontj committed:
Pylint

jacksontj committed:
Fix mis-naming from pylint cleanup

Commit:
Moved tests/unit/loader.py to tests/integration/loader/loader.py

These are integration tests and not unit tests, so they should be
in the proper directory.

Also, fixed up the python 3 incompatability pylint errors.

jacksontj committed:
Add "disabled" support

this allows you to "blacklist" modules.

Addition to #20274

Conflicts:
	salt/loader.py

jacksontj commented:
@thatch45 what does "master change" mean?

thatch45 commented:
it means that someone with serious chops needs to review and manage this PR, like me or mike place kind of people etc.

jacksontj commented:
Ah, gotcha. FYI this was merged into develop in #20274 :)

jacksontj committed:
Correct tag name in loader tests

jfindlay commented:
@jacksontj, it means you are a salt master. :-)

jacksontj committed:
Add support for submodules to LazyLoader

This was a feature that was inadvertently removed in #20274, but since it was undocumented and untested no one noticed. Thanks to @steverweber for pointing it out.

This includes the feature as well as tests.

jacksontj committed:
pylint

jacksontj committed:
Change modulekey for submodule test-- to avoid namespace collisions

jacksontj committed:
Support arbitrarily deep modules

Note: salt can still only *call* modules on the first depth, but they can then import anything below

jacksontj committed:
Add six import to loader

jacksontj commented:
This *should* just be a rebase of those commits to include @rallytime 's move of the tests-- it makes backporting additional features/fixes *much* easier since the tests are where the patch thinks it is.

thatch45 commented on salt/daemons/flo/core.py:
```diff
@@ -536,6 +536,9 @@ def _load_modules(self):
         self.modules.value = salt.loader.minion_mods(self.opts.value)
         self.returners.value = salt.loader.returners(self.opts.value, self.modules.value)
 
+        self.modules.value.clear()
```
I need to look more closely at this one

jacksontj replied:
Actually, those 2 lines are now unnecessary. The original implementation of mine was singletons, but they aren't anymore. Meaning this will create an object and then immediately "clear" it, but it's already empty ;)
